### PR TITLE
Issue 3066: Class text elements translation

### PR DESCRIPTION
**Pull request by jrodriguez-rc**
Added translation for class text elements support

### CODE REVIEW

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 1]
```diff
@@ -133,6 +167,20 @@ CLASS ZCL_ABAPGIT_OBJECT_CLAS IMPLEMENTATION.
       it_lines       = lt_lines
       iv_object_name = lv_object
       iv_language    = mv_language ).
+
+    "<!-- #3066 Class text elements translation
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: Dont add change marks, git keeps track of all changes

**Comment by jrodriguez-rc** [comment]
Comments removed and Travis CI checks passed

**Comment by larshp** [comment]
abaplint config updated, run the pretty printer for the interface

**Comment by jrodriguez-rc** [comment]
Updated

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 2]
```diff
@@ -43,10 +55,28 @@ CLASS ZCL_ABAPGIT_OBJECT_CLAS IMPLEMENTATION.
...
+    DATA: lv_clsname TYPE seoclsname.
+
     super->constructor( is_item     = is_item
                         iv_language = iv_language ).
 
     CREATE OBJECT mi_object_oriented_object_fct TYPE zcl_abapgit_oo_class.
+
+    lv_clsname = is_item-obj_name.
+
+    mv_classpool_name   = cl_oo_classname_service=>get_classpool_name( lv_clsname ).
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: please remove double spaces before the first equal sign

```suggestion
    mv_classpool_name = cl_oo_classname_service=>get_classpool_name( lv_clsname ).
```

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 3]
```diff
@@ -245,6 +308,27 @@ CLASS ZCL_ABAPGIT_OBJECT_CLAS IMPLEMENTATION.
     io_xml->add( iv_name = 'TPOOL'
                  ig_data = add_tpool( lt_tpool ) ).
 
+    IF io_xml->i18n_params( )-serialize_master_lang_only IS INITIAL.
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: perhaps move this part to a separate method?

File: src/zif_abapgit_definitions.intf.abap [hunk 4]
```diff
@@ -364,6 +364,20 @@ INTERFACE zif_abapgit_definitions
     tty_col_spec TYPE STANDARD TABLE OF ty_col_spec
                       WITH NON-UNIQUE KEY tech_name.
 
+  TYPES: tt_langu TYPE STANDARD TABLE OF langu.
```
Comment (larshp) [inline src/zif_abapgit_definitions.intf.abap]: is it possible to place these definitions somewhere else? this interface is too messy already

File: src/objects/zcl_abapgit_object_clas.clas.abap (hunk 3 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_object_clas.clas.abap]: A method to check if only master language is active?

File: src/zif_abapgit_definitions.intf.abap (hunk 4 above)
Comment (jrodriguez-rc) [reply src/zif_abapgit_definitions.intf.abap]: I can pass to zcl_abapgit_objects_super and made a change to read all languages supported, this way we have always (for all object types) system languages

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 5]
```diff
@@ -245,6 +308,27 @@ CLASS ZCL_ABAPGIT_OBJECT_CLAS IMPLEMENTATION.
...
+        ls_i18n_tpool-textpool = add_tpool( lt_tpool ).
+        INSERT ls_i18n_tpool INTO TABLE lt_i18n_tpool.
+
+      ENDLOOP.
+
+      IF lines( lt_i18n_tpool ) > 0.
+        io_xml->add( iv_name = 'I18N_TPOOL'
+                     ig_data = lt_i18n_tpool ).
+      ENDIF.
+    ENDIF.
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: perhaps move this part to a separate method?

File: src/zif_abapgit_definitions.intf.abap (hunk 4 above)
Comment (larshp) [reply src/zif_abapgit_definitions.intf.abap]: I'd like not to change `zcl_abapgit_objects_super`, there is a parallel task looking into overall translation approach

Comment (jrodriguez-rc) [reply src/zif_abapgit_definitions.intf.abap]: What about ZCL_ABAPGIT_OBJECTS_PROGRAM?
Objective of this is always have available languages

File: src/objects/zcl_abapgit_object_clas.clas.abap (hunk 5 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_object_clas.clas.abap]: Maybe do same that DESERIALIZE and separate different object is each method
`  METHOD zif_abapgit_object~serialize_xml.
    serialize_abap(  ).

    serialize_tpool( ).

    serialize_sotr(  ).

    serialize_docu( ).
  ENDMETHOD.`

File: src/zif_abapgit_definitions.intf.abap (hunk 4 above)
Comment (jrodriguez-rc) [reply src/zif_abapgit_definitions.intf.abap]: Created new interface zif_abapgit_lang_definitions

File: src/objects/zcl_abapgit_object_clas.clas.abap (hunk 5 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_object_clas.clas.abap]: Encapsulated in different methods

File: src/zif_abapgit_definitions.intf.abap [hunk 6]
```diff
@@ -25,8 +25,8 @@ INTERFACE zif_abapgit_definitions
   TYPES:
     BEGIN OF ty_file.
       INCLUDE TYPE ty_file_signature.
-  TYPES: data TYPE xstring,
-         END OF ty_file .
+      TYPES: data TYPE xstring,
```
Comment (larshp) [inline src/zif_abapgit_definitions.intf.abap]: try this:
```suggestion
  TYPES: data TYPE xstring,
```

Comment (jrodriguez-rc) [reply src/zif_abapgit_definitions.intf.abap]: I changed indentation for these TYPES, but I think that they were in correct position, they're inside another TYPE

**Comment by larshp** [comment]
sorry, have not had time to look into this change, there are 600+ lines changed, in this very central place, so it has to be done right, ie. spend enough time reviewing and testing

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 7]
```diff
@@ -9,27 +9,68 @@ CLASS zcl_abapgit_object_clas DEFINITION PUBLIC INHERITING FROM zcl_abapgit_obje
...
-      serialize_xml
-        IMPORTING io_xml TYPE REF TO zcl_abapgit_xml_output
-        RAISING   zcx_abapgit_exception.
 
+    DATA mi_object_oriented_object_fct TYPE REF TO zif_abapgit_oo_object_fnc .
+    DATA mv_skip_testclass TYPE abap_bool .
+    DATA mt_langu_additional TYPE zif_abapgit_lang_definitions=>tt_langu .
+    DATA mv_classpool_name TYPE progname .
+
+    METHODS deserialize_abap
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: please move this pretty printing to a separate pull request

File: src/objects/zcl_abapgit_object_clas.clas.abap [hunk 8]
```diff
@@ -39,14 +80,32 @@ ENDCLASS.
...
     CREATE OBJECT mi_object_oriented_object_fct TYPE zcl_abapgit_oo_class.
+
+    lv_clsname = is_item-obj_name.
+
+    mv_classpool_name = cl_oo_classname_service=>get_classpool_name( lv_clsname ).
+
+    " Table d010tinf stores info. on languages in which program is maintained
+    " Select all active translations of program texts
+    " Skip master language - it was already serialized
+    SELECT DISTINCT language
```
Comment (larshp) [inline src/objects/zcl_abapgit_object_clas.clas.abap]: this information is only relevant for serialization? not deserialization and exists? the constructor is executed for all 3 scenarios

File: src/objects/zcl_abapgit_objects_program.clas.abap [hunk 9]
```diff
@@ -782,10 +782,10 @@ CLASS ZCL_ABAPGIT_OBJECTS_PROGRAM IMPLEMENTATION.
                    <ls_field_int>   LIKE LINE OF lt_fieldlist_int.
 
     "#2746: relevant flag values (taken from include MSEUSBIT)
-    CONSTANTS:    lc_flg1ddf TYPE x VALUE '20',
-                  lc_flg3fku TYPE x VALUE '08',
-                  lc_flg3for TYPE x VALUE '04',
-                  lc_flg3fdu TYPE x VALUE '02'.
+    CONSTANTS: lc_flg1ddf TYPE x VALUE '20',
```
Comment (larshp) [inline src/objects/zcl_abapgit_objects_program.clas.abap]: no need to do any changes in this file/no changes to this file, in this case it just makes the PR more difficult to review

File: src/objects/zif_abapgit_oo_object_fnc.intf.abap [hunk 10]
```diff
@@ -1,141 +1,146 @@
...
-      RETURNING
-        VALUE(rt_attributes) TYPE zif_abapgit_definitions=>ty_obj_attribute_tt.
+  TYPES:
+    BEGIN OF ty_includes,
+      programm TYPE programm,
+    END OF ty_includes .
+  TYPES:
+    ty_includes_tt TYPE STANDARD TABLE OF ty_includes WITH DEFAULT KEY .
+
+  METHODS create
```
Comment (larshp) [inline src/objects/zif_abapgit_oo_object_fnc.intf.abap]: pretty printing, please move to separate PR

File: src/objects/zcl_abapgit_object_clas.clas.abap (hunk 7 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_object_clas.clas.abap]: I restore that class to do only minimum changes needed

File: src/objects/zcl_abapgit_object_clas.clas.abap (hunk 8 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_object_clas.clas.abap]: Deleted languages attribute for class definition and move it and selection to method serialize_xml

File: src/objects/zcl_abapgit_objects_program.clas.abap (hunk 9 above)
Comment (jrodriguez-rc) [reply src/objects/zcl_abapgit_objects_program.clas.abap]: You're right, it was changed in a previous commit, I restored it

File: src/objects/zif_abapgit_oo_object_fnc.intf.abap (hunk 10 above)
Comment (jrodriguez-rc) [reply src/objects/zif_abapgit_oo_object_fnc.intf.abap]: I restore that class to do only minimum changes needed

Comment (larshp) [reply src/objects/zif_abapgit_oo_object_fnc.intf.abap]: nice, its now down to around half the size of the original PR 😅

**Comment by larshp** [comment]
thanks, welcome to abapGit as a contributor 🎉 

do note this code will probably be deleted sometime in the future when/if the generic translation approach works
